src/protocols/aggsig/musig_three_rounds.rs

```diff
@@ -0,0 +1,444 @@
+use curv::cryptographic_primitives::commitments::traits::*;
+
+#[derive(Debug)]
+pub struct KeyPair {
+    pub public_key: GE,
+    private_key: FE,
+}
+
+impl KeyPair {
+    pub fn create() -> KeyPair {
+        let ec_point: GE = ECPoint::generator();
+        let private_key: FE = ECScalar::new_random();
+        let public_key = ec_point.scalar_mul(&private_key.get_element());
+        KeyPair {
+            public_key,
+            private_key,
+        }
+    }
+
+    pub fn create_from_private_key(private_key: &BigInt) -> KeyPair {
+        let ec_point: GE = ECPoint::generator();
+        let private_key: FE = ECScalar::from(private_key);
+        let public_key = ec_point.scalar_mul(&private_key.get_element());
+        KeyPair {
+            public_key,
+            private_key,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct KeyAgg {
+    pub apk: GE,
+    pub hash: BigInt,
+}
+
+impl KeyAgg {
+    pub fn key_aggregation(my_pk: &GE, other_pk: &GE) -> KeyAgg {
+        let hash = HSha256::create_hash(&[
+            &BigInt::from(1),
+            &my_pk.bytes_compressed_to_big_int(),
+            &my_pk.bytes_compressed_to_big_int(),
+            &other_pk.bytes_compressed_to_big_int(),
+        ]);
+        let hash_fe: FE = ECScalar::from(&hash);
+        let a1 = my_pk.scalar_mul(&hash_fe.get_element());
+
+        let hash2 = HSha256::create_hash(&[
+            &BigInt::from(1),
+            &other_pk.bytes_compressed_to_big_int(),
+            &my_pk.bytes_compressed_to_big_int(),
+            &other_pk.bytes_compressed_to_big_int(),
+        ]);
+        let hash2_fe: FE = ECScalar::from(&hash2);
+        let a2 = other_pk.scalar_mul(&hash2_fe.get_element());
+        let apk = a2.add_point(&(a1.get_element()));
+        KeyAgg { apk, hash }
+    }
+
+    pub fn key_aggregation_n(pks: &[GE], party_index: usize) -> KeyAgg {
+        let bn_1 = BigInt::from(1);
+        let x_coor_vec: Vec<BigInt> = pks
+            .iter()
+            .map(|pk| pk.bytes_compressed_to_big_int())
+            .collect();
+
+        let hash_vec: Vec<BigInt> = x_coor_vec
+            .iter()
+            .map(|pk| {
+                let mut vec = Vec::new();
+                vec.push(&bn_1);
+                vec.push(pk);
+                for mpz in x_coor_vec.iter().take(pks.len()) {
+                    vec.push(mpz);
+                }
+                HSha256::create_hash(&vec)
+            })
+            .collect();
+
+        let mut apk_vec: Vec<GE> = pks
+            .iter()
+            .zip(&hash_vec)
+            .map(|(pk, hash)| {
+                let hash_t: FE = ECScalar::from(&hash);
+                let pki: GE = pk.clone();
+                pki.scalar_mul(&hash_t.get_element())
+            })
+            .collect();
+
+        let pk1 = apk_vec.remove(0);
+        let sum = apk_vec
+            .iter()
+            .fold(pk1, |acc, pk| acc.add_point(&pk.get_element()));
+
```

Comment:
I guess we can also do 
```
let sum = apk_vec
            .skip(1)
            .iter()
            .fold(apk_vec[0], |acc, pk| acc.add_point(&pk.get_element()));
``` 
without `apk_vec.remove(0)`, and hence without modification on `apk_vec`.

Reply:
it's just a different favor (because it won't be sinificantly different in terms of speed), so i think it doesn't really matter

Reply:
Good idea

Reply:
Good idea